### DISCUSSION TITLE
about indexer question

### DISCUSSION THREAD

**Discussion by Hebe970217** [clangd/clangd#1310] [Q&A]
I don't quite understand the function of indexer.
The description in the document is too short.
I know that it can perform the associated mapping of source files and symbols by compiling the database. 
But is it a component service used only within clangd? 

My goal is to obtain all the symbols in the source file from the compiled database. 
It seems that there is no way for indexer, because it does not contain any interfaces that can be directly obtained. Is there a better way? for example: The library introduced in libclang can directly implement the call interface.

I would be grateful if you could give me an accurate answer
Thank @sam-mccall

**Comment by i-ky** [comment]
> But is it a component service used only within clangd?

I believe this is correct.

> My goal is to obtain all the symbols in the source file from the compiled database.

I think you should use [one of the documented interfaces](https://clang.llvm.org/docs/Tooling.html). I guess [libclang bindings for Python](https://libclang.readthedocs.io/en/latest/) should be sufficient for your use case and relatively easy to use.

**Reply by Hebe970217** [reply]
> > But is it a component service used only within clangd?
> 
> I believe this is correct.
> 
> > My goal is to obtain all the symbols in the source file from the compiled database.
> 
> I think you should use [one of the documented interfaces](https://clang.llvm.org/docs/Tooling.html). I guess [libclang bindings for Python](https://libclang.readthedocs.io/en/latest/) should be sufficient for your use case and relatively easy to use.

Thank you's answer. That true, I used AST iterator parse my cxx source.
But I need to maintain the index from the file association, which will cause the space occupied by my program to become bloated. The indexer will maintain the compiled database, which includes the monitoring of file changes. I think the indexer is very optional.
For a developer, it is really great if you can directly query the language symbols from the Indexer.

**Reply by i-ky** [reply]
If you insist on using indexer, I would recommend to put clangd between your application and indexer. This way clangd will talk to indexer and you don't need to know anything about the protocol they use. Your application will talk to clangd using well-documented [Language Server Protocol](https://microsoft.github.io/language-server-protocol/).

**Reply by Hebe970217** [reply]
Yes, the functional objectives can be achieved through standard protocols

I mentioned before whether clangd supports the acquisition of all 'Tokens' of files, and the answer given to me is sorry. The maintainer mentioned a concept of "strong workspace". As for "strong workspace", I have checked various IDEs, which have different representations. Some have a global symbol tree view, while others do not support it. They just talk about the project root folder path as the workspace, which is exactly consistent with the LSP standard specification.



For example, this goal is achieved by parsing source files in QtCreator. It may also be due to the impact of communication protocol efficiency, and if a single request is kept busy in communication, other functions will not be available.



Or



For example, jdtls (java lsp server) has a framework in which you can query which file a symbol exists in through 'execute/command', but this is not stipulated in the standard protocol.



Of course, more extensive customization may be required for different languages and projects.



Thank you for your answer. I probably know the role of indexer. 
At the same time, I expect indexer to provide more interfaces to facilitate the rapid implementation of various functions
